### PR TITLE
Method `combine` cleans up listeners properly

### PR DESCRIPTION
**Pull request by Starzu**

### CODE REVIEW

File: core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package io.udash.properties.seq
+
+import io.udash.properties.{CrossCollections, PropertyCreator}
+import io.udash.properties.single.{CombinedProperty, ReadableProperty}
+import io.udash.utils.Registration
+
+import scala.collection.mutable
+
+class CombinedReadableSeqProperty[A, B, R: PropertyCreator](
```
Comment (ddworak) [inline core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: If you want to keep this public, add docs.

File: core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  protected var combinedChildren: Option[mutable.Buffer[ReadableProperty[R]]] = None
+  protected var originListenerRegistration: Registration = _
+
+  protected def originStructureListener(originPatch: Patch[ReadableProperty[A]]): Unit = {
+    val combinedNewChildren = originPatch.added.map(sub => sub.combine(p)(combiner))
+    val mappedPatch: Patch[ReadableProperty[R]] = originPatch.copy(
+      removed = originPatch.removed.indices.map(idx => combinedChildren.get(idx + originPatch.idx)),
+      added = combinedNewChildren
+    )
+    CrossCollections.replace(combinedChildren.get, originPatch.idx, originPatch.removed.size, combinedNewChildren:_*)
```
Comment (ddworak) [inline core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: Missing space before varargs expansion

File: core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import scala.collection.mutable
+
+class CombinedReadableSeqProperty[A, B, R: PropertyCreator](
+  s: ReadableSeqProperty[A, _ <: ReadableProperty[A]], p: ReadableProperty[B],
+  combiner: (A, B) => R
+) extends CombinedProperty[Seq[A], B, Seq[R]](s, p, null, (x, y) => x.map(v => combiner(v, y)))
+  with AbstractReadableSeqProperty[R, ReadableProperty[R]] {
+
+  protected var combinedChildren: Option[mutable.Buffer[ReadableProperty[R]]] = None
+  protected var originListenerRegistration: Registration = _
```
Comment (ddworak) [inline core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: Why are these `protected` and not `private`?

File: core/shared/src/main/scala/io/udash/properties/single/CombinedProperty.scala [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import io.udash.properties.MutableBufferRegistration
+import io.udash.utils.Registration
+
+class CombinedProperty[A, B, R](
+  override val origin: ReadableProperty[A], originTwo: ReadableProperty[B],
+  override val parent: ReadableProperty[_], combiner: (A, B) => R
+) extends ForwarderReadableProperty[R] {
+  protected var lastValueOne: Option[A] = None
+  protected var lastValueTwo: Option[B] = None
+  protected var originListenerRegistrations: (Registration, Registration) = _
```
Comment (ddworak) [inline core/shared/src/main/scala/io/udash/properties/single/CombinedProperty.scala]: I believe all 3 could be private.

File: core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala (hunk 2 above)
Comment (Starzu) [reply core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: Did you mean: `:_*` -> `: _*`?

Comment (ddworak) [reply core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: Yeah, that's what IntelliJ would do when reformatting

Comment (Starzu) [reply core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: You saved my life.

Comment (ddworak) [reply core/shared/src/main/scala/io/udash/properties/seq/CombinedReadableSeqProperty.scala]: Always there to safeguard you from eating spaces :rescue_worker_helmet:

File: core/shared/src/main/scala/io/udash/properties/single/CombinedProperty.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    fireValueListeners()
+  }
+
+  protected def originListenerTwo(originValue: B) : Unit = {
+    lastValueTwo = Some(originValue)
+    fireValueListeners()
+  }
+
+  private def initOriginListener(): Unit = {
+    if (originListenerRegistrations == null || !originListenerRegistrations._1.isActive || !originListenerRegistrations._2.isActive) {
```
Comment (FreexD) [inline core/shared/src/main/scala/io/udash/properties/single/CombinedProperty.scala]: please use pattern matching to extract tuple elements

File: core/shared/src/main/scala/io/udash/properties/ImmutableProperty.scala [hunk 6]
```diff
@@ -38,7 +38,7 @@ private[properties] class ImmutableProperty[A](value: A) extends ReadablePropert
   override protected[properties] def fireValueListeners(): Unit = {}
   override protected[properties] def valueChanged(): Unit = {}
   override protected[properties] def validate(): Unit = {}
-  override private[properties] def listenersCount() = 0
+  override def listenersCount() = 0
```
Comment (FreexD) [inline core/shared/src/main/scala/io/udash/properties/ImmutableProperty.scala]: Type annotations are always useful

File: core/shared/src/main/scala/io/udash/properties/single/ReadableProperty.scala [hunk 7]
```diff
@@ -25,7 +25,7 @@ trait ReadableProperty[A] {
   def listenOnce(valueListener: A => Any): Registration
 
   /** Returns listeners count. */
-  private[properties] def listenersCount(): Int
+  def listenersCount(): Int
```
Comment (FreexD) [inline core/shared/src/main/scala/io/udash/properties/single/ReadableProperty.scala]: AbstractReadableProperty is missing override keyword for some methods - including this one. It's harder to find usages of it that way.

File: core/shared/src/main/scala/io/udash/properties/single/ReadableProperty.scala [hunk 8]
```diff
@@ -25,7 +25,7 @@ trait ReadableProperty[A] {
   def listenOnce(valueListener: A => Any): Registration
 
   /** Returns listeners count. */
```
Comment (FreexD) [inline core/shared/src/main/scala/io/udash/properties/single/ReadableProperty.scala]: Isn't this doc to obvious to be repeated in AbstractReadableProperty, captain?

**Pull request merged**